### PR TITLE
reduce the number of I/O operations for HNSW during deserialization

### PR DESCRIPTION
**Pull request by inabao**

### CODE REVIEW

File: src/stream_reader.h [hunk 1]
```diff
@@ -73,6 +81,10 @@ class IOStreamReader : public StreamReader {
     void
     Read(char* data, uint64_t size) override;
 
+    size_t
+    Size() const override;
+
 private:
     std::istream& istream_;
+    size_t max_size_;
```
Comment (wxyucs) [inline src/stream_reader.h]: add default value

File: src/algorithm/hnswlib/hnswalg.cpp [hunk 2]
```diff
@@ -861,50 +862,97 @@ HierarchicalNSW::loadIndex(const std::string& location, SpaceInterface* s, size_
...
 template <typename T>
 static void
-ReadOne(StreamReader& reader, T& value) {
+read_one(StreamReader& reader, T& value, size_t& cursor) {
     reader.Read(reinterpret_cast<char*>(&value), sizeof(value));
+    cursor += sizeof(value);
+}
+
+void
+cache_read_from_block(char* src,
```
Comment (jiaweizone) [inline src/algorithm/hnswlib/hnswalg.cpp]: Porting from other branch ?

**Comment by LHT129** [comment]
I think we don't need to modify so much, just add a fixed size readbuffer allocated by allocator, if read out of the buffer, call the read func to refresh the buffer, otherwise copy data from the buf is ok

File: src/stream_reader.h (hunk 1 above)
Comment (inabao) [reply src/stream_reader.h]: done

File: src/algorithm/hnswlib/hnswalg.cpp [hunk 3]
```diff
@@ -837,24 +837,25 @@ HierarchicalNSW::SerializeImpl(StreamWriter& writer) {
 void
 HierarchicalNSW::loadIndex(std::function<void(uint64_t, uint64_t, void*)> read_func,
                            SpaceInterface* s,
+                           size_t file_size,
                            size_t max_elements_i) {
     int64_t cursor = 0;
-    ReadFuncStreamReader reader(read_func, cursor);
+    ReadFuncStreamReader reader(read_func, cursor, cursor + file_size, allocator_, true);
```
Comment (jiaweizone) [inline src/algorithm/hnswlib/hnswalg.cpp]: minor: for literal constant param we can add desc to improve readability: `ReadFuncStreamReader reader(read_func, cursor, cursor + file_size, allocator_, /*use_buffer=*/true);`

File: src/stream_reader.cpp [hunk 4]
```diff
@@ -14,21 +14,49 @@
 // limitations under the License.
 
 #include "stream_reader.h"
+
+#include <iostream>
```
Comment (jiaweizone) [inline src/stream_reader.cpp]: remove unused header files

File: src/stream_reader.h [hunk 5]
```diff
@@ -23,10 +23,42 @@
 
 class StreamReader {
 public:
-    StreamReader() = default;
+    StreamReader(vsag::Allocator* allocator) : buffer_(allocator) {
```
Comment (jiaweizone) [inline src/stream_reader.h]: buffer_(allocator) looks like `buffer_ = allocator` :)

File: src/stream_reader.cpp (hunk 4 above)
Comment (inabao) [reply src/stream_reader.cpp]: done

File: src/algorithm/hnswlib/hnswalg.cpp (hunk 3 above)
Comment (inabao) [reply src/algorithm/hnswlib/hnswalg.cpp]: I've abstracted out a BufferReader and explicitly used use_buffer.

File: src/stream_reader.h (hunk 5 above)
Comment (inabao) [reply src/stream_reader.h]: This is the invocation of the constructor of buffer_ to initialize the alignment.

File: src/algorithm/hnswlib/hnswalg.cpp [hunk 6]
```diff
@@ -920,10 +903,12 @@ HierarchicalNSW::DeserializeImpl(StreamReader& reader, SpaceInterface* s, size_t
                 throw std::runtime_error(
                     "Not enough memory: loadIndex failed to allocate linklist");
             reader.Read(link_lists_[i], link_list_size);
+            cursor += link_list_size;
         }
     }
     if (normalize_) {
         reader.Read(reinterpret_cast<char*>(molds_), max_elements_ * sizeof(float));
+        cursor += max_elements_ * sizeof(float);
```
Comment (jiaweizone) [inline src/algorithm/hnswlib/hnswalg.cpp]: cursor not use ?

File: src/footer.cpp [hunk 7]
```diff
@@ -15,6 +15,7 @@
 
 #include "footer.h"
 
+#include <iostream>
```
Comment (jiaweizone) [inline src/footer.cpp]: remove unused header file

File: src/impl/conjugate_graph.cpp [hunk 8]
```diff
@@ -15,6 +15,7 @@
 
 #include "conjugate_graph.h"
 
+#include <iostream>
```
Comment (jiaweizone) [inline src/impl/conjugate_graph.cpp]: ditto

File: src/algorithm/hnswlib/hnswalg.cpp (hunk 6 above)
Comment (inabao) [reply src/algorithm/hnswlib/hnswalg.cpp]: I have removed all the unused modification.

File: src/footer.cpp (hunk 7 above)
Comment (inabao) [reply src/footer.cpp]: I have removed it

File: src/impl/conjugate_graph.cpp (hunk 8 above)
Comment (inabao) [reply src/impl/conjugate_graph.cpp]: I have removed it

**Review by jiaweizone — APPROVED** [review]
LGTM

**Review by LHT129 — APPROVED** [review]
LGTM

**Pull request merged**